src/pharmpy/expressions.py

```diff
@@ -56,3 +56,142 @@ def _free_images_iter(expr: sympy.Expr) -> Iterable[sympy.Expr]:
+        else sympy.sympify(old, strict=not isinstance(old, type))
+    )
+
+
+@lru_cache(maxsize=256)
+def _sympify_new(new) -> sympy.Expr:
+    # NOTE This mimics sympy's input coercion in subs
+    return sympy.sympify(new, strict=not isinstance(new, (str, type)))
+
+
+def subs(expr: sympy.Expr, mapping: Dict[sympy.Expr, sympy.Expr], simultaneous: bool = False):
+    _mapping = xreplace_dict(mapping)
+    if (simultaneous or _mapping_is_not_recursive(_mapping)) and all(
+        map(_old_does_not_need_generic_subs, _mapping.keys())
+    ):
+        if sympy.exp in _mapping:
+            new_base = _mapping[sympy.exp]
+            _mapping[sympy.exp] = lambda x: new_base**x
+            _mapping[sympy.Pow] = lambda a, x: new_base**x if a is sympy.S.Exp1 else a**x
+            return _subs_atoms_simultaneously(_subs_atom_or_func(_mapping), expr)
+        return _subs_atoms_simultaneously(_subs_atom(_mapping), expr)
+    return expr.subs(_mapping, simultaneous=simultaneous)
+
+
+def _mapping_is_not_recursive(mapping: Dict[sympy.Expr, sympy.Expr]):
+    return set(mapping.keys()).isdisjoint(
+        set().union(*map(lambda e: e.free_symbols, mapping.values()))
+    )
+
+
+def _old_does_not_need_generic_subs(expr: sympy.Expr):
+    return isinstance(expr, sympy.Symbol) or expr is sympy.exp
+
+
+def _subs_atoms_simultaneously(
+    subs_new_args: Callable[[sympy.Expr, List[sympy.Expr]], sympy.Expr], expr: sympy.Expr
+):
+    stack = [expr]
+    output = [[], []]
+
+    while stack:
+
+        e = stack[-1]
+        old_args = e.args
+
+        new_args = output[-1]
+
+        n = len(old_args)
+        i = len(new_args)
+
+        if i == n:
+            stack.pop()
+            output.pop()
+            output[-1].append(subs_new_args(e, new_args))
+        else:
+            # NOTE Push the next argument on the stack
+            stack.append(old_args[i])
+            output.append([])
+
+    return output[0][0]
+
+
+def _build(expr: sympy.Expr, args: List[sympy.Expr]):
+    return expr if all(map(is_, expr.args, args)) else expr.func(*args)
+
+
+def _subs_atom(mapping: Dict[sympy.Expr, sympy.Expr]):
+    def _subs(expr: sympy.Expr, args: List[sympy.Expr]):
+        return _build(expr, args) if args else mapping.get(expr, expr)
+
+    return _subs
+
+
+def _subs_atom_or_func(mapping: Dict[sympy.Expr, sympy.Expr]):
+    def _subs(expr: sympy.Expr, args: List[sympy.Expr]):
+        if not args:
+            return mapping.get(expr, expr)
+
+        fn = expr.func
+        new_fn = mapping.get(fn, fn)
+        return _build(expr, args) if fn is new_fn else new_fn(*args)
+
+    return _subs
+
+
+def _neutral(fn: sympy.Function) -> sympy.Integer:
+    if fn is sympy.Add:
+        return sympy.Integer(0)
+    if fn is sympy.Mul:
+        return sympy.Integer(1)
+    if fn is sympy.Pow:
+        return sympy.Integer(1)
+
+    raise ValueError(f'{type(fn)}: {repr(fn)}')
+
+
+def prune(predicate: Callable[[sympy.Expr], bool], expr: sympy.Expr):
```

Comment:
What does this function do? Perhaps a small comment to explain.

Reply:
It removes nodes from expressions. I use it instead of `expr.subs({sympy.exp: 1})` which replaces `e^x` subexpressions by `1`. I'll add a docstring.

Reply:
> I'll add a docstring.

Done.